api/include/opentelemetry/context/context.h

```diff
@@ -0,0 +1,205 @@
+  // be the head of the context object linked list.
+  template <class T , nostd::enable_if_t<trace::detail::is_key_value_iterable<T>::value> * = nullptr>
+  Context(const T &keys_and_values)  
+  {
+    head_ = BuildList(keys_and_values);
+    next_ = NULL;
+  }
+  
+  // Creates a context object from a key and value, this will be the head
+  // of the context object linked list.
+  Context(nostd::string_view key, ContextValue value)  
+  {
+    head_ = BuildHead(key, value);
+    next_ = NULL;
+  }
+
+  // Accepts a new iterable and then returns a new context that
+  // contains the new key and value data.
+  template <class T , nostd::enable_if_t<trace::detail::is_key_value_iterable<T>::value> * = nullptr>
+  Context* SetValues(T &values) noexcept
+  {
+    Context *new_node = new Context(values, this);
+    return new_node;
+  }
+  
+  // Accepts a new iterable and then returns a new context that
+  // contains the new key and value data.
+  Context* SetValue(nostd::string_view key, ContextValue value) noexcept
+  {
+    Context *new_node = new Context(key,value, this);
+    return new_node;
+  }
+
+  // Returns the value associated with the passed in key.
+  common::AttributeValue GetValue(nostd::string_view key)
+  {
+    Context *context_trace = this;
+    
+    //Iterate through the context nodes
+    while(context_trace != NULL){
+      DataList *data_trace = context_trace->head_;
+      
+      //Iterate through the internal data nodes
+      while(data_trace != NULL){
+        if(strncmp(key.data(), data_trace->key_, 50) == 0){
+          return data_trace->value_;
+        }
+        
+        data_trace = data_trace->next_;
+      }
+      
+      context_trace = context_trace->next_;
+    }
+    
+    return "";
+  }
+
+  // Iterates through the current and comparing context objects
+  // to check for equality.
+  bool operator==(const Context &other)
+  {
+    // If the two contexts are not at the place in the list
+    // they cannot be equivalent
+    if(next_ != other.next_){
+      return false; 
+    }
+    
+    DataList *trace = head_;
+    DataList *other_trace = other.head_;
+    
+    // Iterate through the lists and compare the keys
+    // TODO should also compare the values but the nostd::variant
+    // seems to not have a comparison operator
+    while(trace != NULL){
+
+      if(other_trace == NULL){
+        return false;
+      }
+      
+      if((trace->key_ != other_trace->key_)){
+        return false;
+      }
+
+      trace = trace->next_;
+      other_trace = other_trace->next_;
+    }
+
+    return true;
+  }
+
+  Context(const Context &other) = default;
+  
+  Context &operator=(const Context &other){
+    head_ = other.head_;
+    next_ = other.next_;
+  }
+
+private:
+  
+  // Head of the list which holds the keys and values of this context 
+  DataList *head_;
+  
+  // Pointer to the next context object in the context list
+  Context *next_;
```

Comment:
Consider setting next_ to nullptr here and not in each of the constructors.

Reply:
Set to nullptr in class def.